source/cl/source/extension/source/intel_unified_shared_memory/intel_unified_shared_memory.cpp

```diff
@@ -418,6 +424,141 @@ cl_int device_allocation_info::allocate(cl_uint alignment) {
+    : allocation_info(context, size),
+      device(device),
+      mux_memory(nullptr),
+      mux_buffer(nullptr) {
+  if (device) {
+    cl::retainInternal(device);
+  }
+};
+
+shared_allocation_info::~shared_allocation_info() {
+  if (mux_buffer) {
+    assert(device);
+    (void)muxDestroyBuffer(device->mux_device, mux_buffer,
+                           device->mux_allocator);
+  }
+
+  if (mux_memory) {
+    assert(device);
+    muxFreeMemory(device->mux_device, mux_memory, device->mux_allocator);
+  }
+
+  // Free the host side allocation
+  if (base_ptr) {
+    cargo::free(base_ptr);
+  }
+
+  if (device) {
+    cl::releaseInternal(device);
+  }
+};
+
+cargo::expected<std::unique_ptr<shared_allocation_info>, cl_int>
+shared_allocation_info::create(cl_context context, cl_device_id device,
+                               const cl_mem_properties_intel *properties,
+                               size_t size, cl_uint alignment) {
+  OCL_CHECK(size == 0, return cargo::make_unexpected(CL_INVALID_BUFFER_SIZE));
+
+  OCL_CHECK((alignment & (alignment - 1)) != 0,
+            return cargo::make_unexpected(CL_INVALID_VALUE));
+
+  if (device) {
+    const auto device_align = device->mux_device->info->buffer_alignment;
+    OCL_CHECK(size > device->max_mem_alloc_size,
+              return cargo::make_unexpected(CL_INVALID_BUFFER_SIZE));
+    OCL_CHECK(alignment > device_align,
+              return cargo::make_unexpected(CL_INVALID_VALUE));
+
+    if (alignment == 0) {
+      alignment = device_align;
+    }
+  }
+
+  auto alloc_properties = parseProperties(properties, true);
+  if (!alloc_properties) {
+    return cargo::make_unexpected(alloc_properties.error());
+  }
+
+  // Default alignment with no device is the highest alignment of all other
+  // devices
+  if (!device && alignment == 0) {
+    cl_uint max_align = 1;
+    for (auto device : context->devices) {
+      const auto device_align = device->min_data_type_align_size;
+      max_align = device_align > max_align ? device_align : max_align;
+    }
+    alignment = max_align;
+  }
+
+  auto usm_alloc = std::unique_ptr<shared_allocation_info>(
+      new (std::nothrow) shared_allocation_info(context, device, size));
+
+  // Decide whether to allocate initially on device or not
+  // NOTE: The specification says these are only hints, so we ignore them for
+  // now
+  bool prefers_host = true;
+  if (*alloc_properties & CL_MEM_ALLOC_INITIAL_PLACEMENT_DEVICE_INTEL) {
+    prefers_host = false;
+  }
+  (void)prefers_host;
+
+  cl_int error = usm_alloc->allocate(alignment);
+  OCL_CHECK(error != CL_SUCCESS, return cargo::make_unexpected(error));
+
+  usm_alloc->alloc_flags = alloc_properties.value();
+
+#if defined(__GNUC__) && !defined(__clang__) && __GNUC__ < 9
+  // GCC <9 requires this redundant move, this branch of the #if can be
+  // deleted once the minimum supported version of GCC is at least 9.
+  return std::move(usm_alloc);
```

Comment:
Could you give a description of the bug that you are working around here (can be a code snippet here, a GCC bug report link, a Godbolt link, whatever is most convenient) I tried to figure it out by experimenting on Godbolt, but even with non-movable types, even with a mismatch between local variable type and return type, GCC accepted my attempts before GCC 9 as well, so there is something more to it.

Reply:
Created internal issue for this - it's in multiple places and we want to see if it's still needed.